zerver/tests/test_events.py

```diff
@@ -555,14 +555,74 @@ def test_send_message_to_existing_recipient(self) -> None:
             state_change_expected=True,
         )
 
-    def test_add_reaction(self) -> None:
-        message_id = self.send_stream_message(self.example_user("hamlet"), "Verona", "hello")
+    def test_add_and_remove_reaction_stream(self) -> None:
```

Comment:
This test will likely want a lot more data setup to properly verify the matrix of cases that we want to consider:
* private message
* group private message
* public stream, guest user with UserMessage not subscribed
* public stream, guest user without UserMessage not subscribed
* public stream, guest user with UserMessage subscribed
* public stream, guest user without UserMessage subscribed
* That same batch for normal user

I think the right way to do this is to just write a new test from scratch, start with having Iago create a new test stream, send a message_before_subscribe, subscribe Polonius and Hamlet, and then send message_after_subscribe, and then have Iago react to both messages and verify the set of users that were notified for each reaction.  And then change the settings for the stream's privacy and repeat with a different emoji.

Reply:
Oh, also, I'd add the new test outside `test_events.py` -- this test suite is mainly for testing a different part of the system.  `test_reactions.py` with careful use of `with tornado_redirected_to_list` to capture the set of notified user IDs will be a lot happier than the loops here.